### DISCUSSION TITLE
Split tunneling based on domain names/IP addresses

### DISCUSSION THREAD

**Discussion by jorik392** [mullvad/mullvadvpn-app#9059] [Feature requests & Ideas]
Split tunneling is currently limited to specific applications. 

It will be useful to also have the ability to do this for specific domain names/IP addresses so traffic to them will not go through the VPN.

**Comment by faern** [comment]
Before starting the split tunneling work we evaluated doing it per IP/IP range or per application. We decided that per application was the most useful of the two for the majority of users. So we went with that.

There are currently no plans for other types of split tunneling. That does not mean it will never exist, just that we are currently not planning for it or have a timeline for it.

**Comment by jorik392** [comment]
Can this be reopened if you think this has a chance to be implemented in the future? Many other VPN software including the generic OpenVPN client offer the ability to do this and it would be beneficial to users that have no choice but to disable the VPN for certain domain names/IP addresses.

**Comment by Hurstwood** [comment]
Jorik392, if you are thinking of websites like legitimate streaming services that do not permit access via a VPN/proxy, you could always have a dedicate browser for non-VPN traffic. 

Chrome, Firefox, and Edge all have stable and nightly releases.  That's 6 available browsers that doesn't include alpha and beta releases.  I'm not sure about Safari; and then there are other browsers but that depends on how you feel about them. 

For me, i prefer internet traffic to be routed through the VPN by default and specific internet traffic to be routed outside of the VPN by making specific rules - a white list.  Looking at Openvpn's documents, it appears to function the opposite way.  Internet traffic is routed outside of the VPN by default and you create rules for traffic to be routed through it.  Remember, Openvpn is meant to provide a user access to a private network, not obfuscate their activity.

An issue i see with split tunneling via domain name or IP address is that most websites pull resources from several servers, therefore for one website you will have to create several rules.  Every server can be accessed via several IP addresses.  What happens when a server changes its IP addresses or resolved name - a user will think they're protected when they are not.  It sounds like a disaster waiting to happen.

**Comment by raphpa** [comment]
Split tunneling for IP addresses also has other use cases and is implemented in other VPN clients like PIAs.

Two use cases I have right now:
- My machine is on the 192.168.0.0/24 subnet and I (intentionally) have another local subnet 192.168.100.0/24 which is accessible through my router and its firewall. I cannot reach that subnet right now as the VPN client only seems to whitelist the local subnet of my current machine.

- Tailscale uses the 100.64.0.0/10 subnet, so you need to be able to whitelist this block in order to use Tailscale at the same time as the mullvad VPN to reach your servers.

**Comment by faern** [comment]
You should be able to reach `192.168.100.0/24` if you enable local network sharing. However, mind that you need to have a route set up for this network. It has to be a route more specific than the default route that absorbs all traffic to the tunnel, then it will work. But that's probably what you should have anyway if you intend to communicate with that network.

Depending on your operating system you can always set up your own custom split tunneling with the firewall/routing table. If you use Linux, see this guide for some tips and tricks to exclude whatever types of packets: https://mullvad.net/en/help/split-tunneling-with-linux-advanced/

**Comment by ghost** [comment]
Another good purpose of this would be w/ IP locked services, which don't let you connect from a different IP address. Since Mullvad won't let you have dedicated IPs, it's pretty much just constantly requesting to change your IP address or not using a VPN at all, at least that's how it went in my case

**Comment by Drakonas** [comment]
Is this possible to be implemented? It would be really handy. Some private websites don't allow you to use a VPN to browse them (and ban when you do so), but allow it to be used with connections through other apps.

Having an IP/domain whitelist would be very handy in these use cases, because otherwise you have to exclude the browser entirely instead of only the specific site you need excluded (and whatever connections it uses).

**Comment by faern** [comment]
@Drakonas The current recommendation is to run multiple browsers. One that is excluded and one that is not. Then you can do your non-VPN stuff in the excluded one.

**Comment by relativisticelectron** [comment]
It would be great if this nftables based method would be integrated into the Linux version: https://github.com/relativisticelectron/python-mullvad-tailscale

**Comment by trollollol123** [comment]
Implementing this could solve the voice chat issues people have in games like valorant and overwatch?

**Comment by galennare** [comment]
Additionally would resolve common issues people might have with work VPN/secure virtual desktops people might use for remote work.

**Comment by JGKle** [comment]
Just trying out Mullvad for the first time, this missing feature is an immediate headache:

1. I have to use Twingate for work, but it can't connect as long as I'm on Mullvad (same issue as Tailscale mentioned above)
2. I use multiple sites that won't let me access them through a VPN. I know I can use an entirely separate browser, but that's a hassle, and means I need to do that on every system - 2 dual-boot PCs and 1 phone = 5 OSs. That's a lot of extra browsers just to circumvent this. I really like Mullvad's privacy, but this adds a lot more friction than expected - figured I could just easily exclude traffic I don't want to run through the VPN...

**Comment by faern** [comment]
> 1. I have to use Twingate for work, ...

That's unfortunate :( For compatibility with separate VPN-isch products I mainly recommend contacting our support, since this is a feature request. And/or maybe they (support) know of a workaround or similar.

> 2. I use multiple sites that won't let me access them through a VPN. I know I can use an entirely separate browser, but ...

Sounds like you run Mullvad on some router or something? Or how is IP split tunneling solving the issue on five devices with just one split? If you run Mullvad separately on all five devices you still need to set up the splitting five times over. I'd argue that this is not that much less work than installing an extra browser on five devices.

The current stance is that for splitting specific websites we recommend a separate browser.

**Comment by JGKle** [comment]
> For compatibility with separate VPN-isch products I mainly recommend contacting our support, since this is a feature request. And/or maybe they (support) know of a workaround or similar.

Yeah I already did. They refused to help.

> Sounds like you run Mullvad on some router or something?

No...? Just using it on my PC.

> Or how is IP split tunneling solving the issue on five devices with just one split?

...?? It's not.

> If you run Mullvad separately on all five devices you still need to set up the splitting five times over. I'd argue that this is not that much less work than installing an extra browser on five devices.

The one-time setup is similar, yes. But that's just one-time. Now on an ongoing basis, you need to have twice as many browsers installed everywhere. Twice as many browsers using up storage on each OS, twice as many browsers needing to be kept up to date, needing to quit & launch another separate process rather than simply alt+T a new tab, and so on and so on. Having to remember on an ongoing basis which sites you need to use through a separate browser. It's far, far more user friction & a far worse user experience on an ongoing basis vs just doing the up-front setup & then everything working seamlessly & the same: getting to use one browser for everything like normal.

**Comment by faern** [comment]
There are multiple problems with splitting the IP for example.com and then just hoping it works well.

1. example.com might change IP. Requiring something to monitor the domain and check for changes. Also larger services have a huge amount of associated IPs and they might return random replies at different times, making it hopeless to enumerate all IPs for a given domain.
2. When you load one website it's not talking strictly to one domain. It will load truckloads of resources from various places, usually. So when you load the excluded-from-tunnel example.com should the dependency cdn.some-host.xyz be loaded in the tunnel or outside the tunnel? If another-example.com that is not excluded also loads from cdn.some-host.xyz should it be different? If the splitting happens on an IP level, it can't really. And this might impact the end result on how example.com behave
3. Many services can share a single IP. A single IP can host both example.com and another-example.com. The server knows what page you want to load from the `host` header in your request. This means that if you exclude the IP for example.com you will also indirectly exclude any traffic to other services sharing the same IP.

All of the above combined makes it really hard to reason about behavior, security and privacy. Thus we don't really recommend tunnel splitting based on IP, but rather per application.

Our own Firefox browser extension, and our own browser supports per-domain exit servers. Meaning you can custom route some websites to a different VPN server. This is sort of what you want! Except that you can't tell it to route anything outside the tunnel altogether. Having an option to route things outside the tunnel altogether would not be technically impossible. However, it would require some deeper integration between our app and our browser extension. This has currently not been prioritized.

**Comment by JGKle** [comment]
Understood, thanks for that context.

Honestly at this point my real blocker is the Twingate thing anyway, since the website thing at least does have a workaround. For Twingate, I simply have to disconnect from Mullvad entirely - otherwise I'm unable to work. The result has been that...I'm pretty much never able to use Mullvad at all.

**Comment by bak1an** [comment]
Same as above, I need to allow certain network to go through other vpn for work, mullvad is very aggressive and blocks everything. 
Local network sharing is sort of a solution but I need just one small /24 network, not all possible local networks to be open.

**Reply by AutonomousCat** [reply]
This is a great example of how the addition of this feature could improve security

**Comment by JGKle** [comment]
For what it's worth, I ultimately gave up on using Mullvad. Since my work VPN is non-optional I couldn't really connect to Mullvad the vast majority of the time. After lots of reading I'm certain there's a way to for them to work together, but both ends basically just said I had to ask the other, and I didn't have further time to dedicate to trying to figure it out myself. So I had to drop Mullvad :(

**Comment by bak1an** [comment]
Based on https://github.com/mullvad/mullvadvpn-app/issues/2071#issuecomment-1212210522 and https://mullvad.net/en/help/split-tunneling-with-linux-advanced but for macos I did `pfctl` wrapper to allow my specific network - https://gist.github.com/bak1an/33a920627d70d6d8c08bba362c83369a

Now it works; route was already present from other vpn, only thing blocking it was mullvad's PF rules.

Had to stuff it under `apple.com/` anchor so it gets executed before other mullvad's rules.

It would be great if mullvad's default PF ruleset also had an empty table, like `mullvad_allow` and pass rules for addresses from it.

Then the whole custom routing on mac os would be a simple:
1) Set the route to custom network (or have it set by other vpn for you) 
2) Add this network to this PF table

Much better than figuring out all the mess apple did with PF and anchors and whatnot.

**Comment by jm355** [comment]
This would also be nice to have to work around #7195, especially if you could only exclude a url for certain wifi networks

**Comment by quantenzitrone** [comment]
My university has some services that are only reachable via the universities vpn, but I can't connect to the universities vpn if I am connected to Mullvad. So every time I want to use these services, I have to disconnect my whole device from Mullvad, so I can connect to the universities vpn.
While I could work around that by running everything form my university in a container/vm and split tunnel that container/vm, I would really rather exempt my universities IP address range from mullvad.

**Comment by meichengg** [comment]
I think i got solution on macOS (i'm not quite sure about the technical stuffs behind but it worked for my case). I'm a macOS user so i don't get to test on other platforms.

I'm want to run both VPNs at the same time.

1. Wireguard VPN config running with Wireguard UI app for internal access.
2. Mullvad VPN for public access.

What i did was running both VPN - i wanna make sure, so run the internal access first then Mullvad:
1. I connected to my Wireguard VPN server using the app mentioned.
2. Then i connected to Mullvad VPN.
3. Then i ran this command `sudo pfctl -d && sudo pfctl -f /etc/pf.conf && sudo pfctl -e` and the Wireguard VPN wasn't going through Mullvad like it was.

I also tried with Tailscale (my homelab) + that Wireguard VPN running + Mullvad VPN at the same time just by running that command.

**Note for Tailscale users:**
You have to turn off every DNS filter in Mullvad in order to not mess with Tailscale Magic DNS. If anyone manages to have Mullvad DNS filters enabled and can run parallel with Tailscale, please let me know.

Oh, I didn't use split-tunneling feature of Mullvad in any case! If you guys having clear observation about the technical behind, please help me to explain. Much appreciated!

**Comment by uponminiature** [comment]
Amnezia's implementation of split-tunneling is nearly there and Mullvad could likely take inspiration from it. Has two separate sections: "Site-based" and "App-based" split tunneling, and in the site-based section, you can specify domains as well as IPv4 CIDRs which has allowed me to use Tailscale alongside it (simply by excluding 100.64.0.0/10). It's nearly there because it seems you cannot specify IPv6 CIDRs yet. https://docs.amnezia.org/documentation/instructions/vpn-split-tunneling

**Comment by AutonomousCat** [comment]
I'm surprised this is still being discussed after 5 years. Other well known VPN options support IP/domain based split tunneling.

**Comment by musabagriyanik** [comment]
So it's 5 years later still only a discussion, if you ask me, this is now a mandatory feature. I can give one example, due to the limitations of Java it's not possible to split tunnel java applications, so a Minecraft Java player can't split tunnel, thus, they might have to play with higher latency.

If there was an IP based split tunneling alongside applications, the user could just add that servers IP address to exclusion list, and play with less latency. That was a workaround I found when I was using Proton VPN, because they allowed me to exclude IPs from VPN tunnel. Split tunneling java apps was impossible there as well. Even better, Windscribe has domain split tunneling, so you can just split tunnel the servers domain, even if their IP changes, you'll still be split tunneled. So why Mullvad shouldn't do it?

I don't request this feature for browsers, because, from what I learned when I was using (more like testing for a month) Windscribe, it's best to have two browsers and exclude one of them, that you'll access websites without VPN. Because domain based split tunneling doesn't work well with browsers, due to the nature of web, they made request to multiple other domains/subdomains that might also block VPNs, so you may end up excluding tens of domains to access one website. So duo-browser is better there. I request this feature for exactly the reasons above, there are cases where when we access an IP/domain, we really mean to access one IP/domain. Game servers, or maybe your selfhosted server that you blocked access from some IP ranges? Are you going to split tunnel ssh every time?

**Comment by AutonomousCat** [comment]
Has anyone been successful in modifying the code in the latest versions to allow an IP range? I tried adding 100.64.0.0/10 to ALLOWED_LAN_NETS in an attempt to get Tailscale working alongside Mullvad's own app, and while I was able to connect over Tailscale successfully, the internet outside of Tailscale stopped working 🙃.

**Reply by musabagriyanik** [reply]
> Has anyone been successful in modifying the code in the latest versions to allow an IP range? I tried adding 100.64.0.0/10 to ALLOWED_LAN_NETS in an attempt to get Tailscale working alongside Mullvad's own app, and while I was able to connect over Tailscale successfully, the internet outside of Tailscale stopped working 🙃.

Almost a year ago I tried this, and it didn't work back then unfortunately. I had to add that IP address to multiple parts of the code to make it work, I don't know Rust much, but the code was already there for other subnets so I copied them. However, even when I managed to connect Tailscale and maintain system-wide Mullvad connection, the Tailscale connection was through a [relay](https://tailscale.com/kb/1591/peer-relays), and that makes it incredibly useless. Maybe I did something wrong, I don't know, but I lost hope making it work myself.

**Comment by epoch-philosophy** [comment]
Man, I am genuinely surprised this is still only a discussion. Hands down, this is incredibly useful and important for people who self-host software in their own network using specific domains.

Please do consider implementing this. Until then, I will have to use another solution!

**Comment by AutonomousCat** [comment]
For Tailscale use,  I've resorted to just closing Mullvad and running the native [Wireguard](https://www.wireguard.com/install/) application that uses Mullvad's [configuration](https://mullvad.net/en/account/wireguard-config) with the IP range of my router and Tailscale thrown in (I used some IP calculator to help). 

It's not only a hassle, but it could potentially expose your IP outside of the context of Tailscale if you don't configure it right or don't immediately switch between VPN apps. I probably already do expose my IP in that brief inbetween moment considering background telemetry, auto updaters, etc.

If being "unsafe" is Mullvad's reason for not adding this feature, then it'll do more good than bad to add it natively than to make people find workarounds.

**Comment by uponminiature** [comment]
I'd be using Mullvad all the time via auto connect but I can't because software on my computer which rely on services behind Tailscale stop functioning, like Bitwarden and Nextcloud. It's tedious to disconnect Mullvad and connect Tailscale every time I need to do something which requires a service behind Tailscale - so I just don't use Mullvad for most occasions anymore.

**Comment by musabagriyanik** [comment]
<img width="644" height="272" alt="image" src="https://github.com/user-attachments/assets/87d02419-11d6-4af0-bc19-568a760583e8" />
I get this error in Gemini when I'm connected to Turkey servers, because -based on my research- a few months ago Gemini has changed their region detection system so it's thinking that I'm in Iran, which is where Gemini isn't available. Thus, I get this product is not supported in your country text, and sometimes I get this 403 error page. I can also see this happening when I do search in Google, it finds search results with Arabic letters. I'm all in latin, and everything set to Turkey, except the location setting in Google search engine setting, it's set to current location. So it thinks I'm in Iran or an Arabic country, unless I go into search settings and switch location to Turkey manually.

I want to keep connecting Turkey servers, but I also want to use my Gemini Pro plan, thus, everytime I need it, I need to open my second browser, which is added to split tunnel.

Only if there was domain exclusion I wouldn't have to deal with all this.

**Reply by musabagriyanik** [reply]
Mullvad is only a cencorship circumvention tool for me, not privacy.

**Comment by fiso64** [comment]
I hope this gets implemented. I am currently forced to choose between privacy (VPN DNS) and local functionality (homeserver).

I self-host services at `*.my-domain.com` which resolve to local IPs (`192.168.x.x`) when I am at home (using OpenWrt/Dnsmasq).
When Mullvad is active, it intercepts all DNS queries. If I use "Local Network Sharing", I can ping the server IP, but I lose the DNS resolution for my local domains (Green Padlock breaks).
Domain-based Split DNS would allow me to route `*.my-domain.com` to my local gateway while keeping all other traffic secure inside the tunnel. This is standard in enterprise VPNs and is the only thing missing here.

**Comment by kenneth-ge** [comment]
I need this feature as well, since I need to do development on an IP-locked web server and it does not make sense for me to do this on a per-application basis. At the very least, it would be nice if split-tunneling could also support a whitelist mode, where we could specify which apps go through the VPN instead of which ones don't.

**Comment by carlos-ramos-73** [comment]
Documenting my vote here, +1.
Many issues and discussions throughout the repo point to this same feature.
I understand the official decision of keeping it app based.
However, my assumption is that many of us customers are indeed just technical nerds.
Yes, I will be building my own fork with IP changes. I would rather not.
I just feels like an extra button with a warning inside the "Split tunneling" menu could suffice.

**Comment by RealHypnoticOcelot** [comment]
+1!
I use browser profiles in order to constrain my browser usage/history to different contexts. If I launch my browser with Split Tunneling enabled, and then launch a profile, the new window that opens with the profile won't bypass the VPN. Thus, any non-default profiles can't bypass the VPN! I'd like to just be able to whitelist the websites I frequent that don't allow VPNs.

**Comment by incendius679** [comment]
I agree with this functionality, for work, only ONE website don't work, and I don't want to use another browser just for one website. I really want it, it would be useful.

**Comment by Paul-469** [comment]
I really think this needs to be implemented. Especially, for situations where multiple services run over the same domain. While installing another browser might not seem like that much effort, though I would still argue it is extremely annoying and invites user error by accidentally using the wrong instance, it becomes more annoying the more runs over the domain. Especially universities do that.

If I have to install a second browser, mail client, IDE, SSH client, and what not. Just to have another thing not work because it want to fetch data from that domain you end up with very long white list and you might as well dual boot or spin up a VM at this point.

This is IMO one of the best QoL feature requests and besides that, having a non tunnelling version of browser etc. is asking for user error.

**Comment by Phayderz** [comment]
This should SERIOUSLY be implemented!

**Comment by musabagriyanik** [comment]
Looks like they dgaf

**Comment by no-usernames-left** [comment]
There can very simply be a switch to exclude all traffic with a destination address of RFC1918 netblocks (and the equivalent for IPv6).

**Comment by NeutralKaon** [comment]
Found this feature request upon looking through lots of bugs. I'd use this if it were here. Lord knows it's easier than monkeying around with nft, which is the other option…
